API: add resultSchema() method to StructTransform

stevenzwu opened the pull request:
This can be used to build `Comparator` for `SortKey`, which contains the transformed view of the original struct. `SortOrderComparators` only works with the original struct.

jbonofre commented:
@stevenzwu do we have a corresponding issue for this PR ? Your PR looks like a good idea to me.

stevenzwu commented:
> @stevenzwu do we have a corresponding issue for this PR ? Your PR looks like a good idea to me.

@jbonofre thanks for taking a look. I don't think Github issue is required for every PR in Iceberg. Were you looking for more context than the PR description?

szehon-ho commented:
@stevenzwu Yes actually would you mind you give a little more context ?  Not very familiar with this.

jbonofre commented:
@stevenzwu agree to not create a GH Issue, but my ask is more to get more context on this change. At first glance, it looks good but I would like to have a better understanding about problem to solve or new features added.
I guess it's to extend `StructTransform` with new functionality.

So, yes, I would love to have more context.

stevenzwu commented:
We added the `StructTransform` base class in PR #7798 so that `SortKey` can reuse the code from old `PartitionKey`. `SortKey` was added to support the Flink range distribution implementation:  https://github.com/apache/iceberg/projects/27. 

We added `SortOrderComparators` in PR #7798. It can be used to compare the original row (full schema) with the projected view from `SortOrder`. But during the Flink range distribution implementation, we need to compare the `SortKey` with the `resultSchema` from `StructTransform` projected view using the struct `Comparators` . Hence we need the `StructTransform` to expose the result schema to construct `Comparators`

szehon-ho commented on api/src/main/java/org/apache/iceberg/StructTransform.java:
```diff
@@ -125,6 +148,18 @@ public int hashCode() {
+   * The transform result type is useful for building a comparator for the transformed struct (like
+   * {@link SortKey} using {@link Comparators#forType(Types.StructType)}.
+   *
+   * <p>
```
Nit: do we need this new paragraph?

szehon-ho reviewed (approved):
This looks reasonable to me, just some comments on the javadoc side

stevenzwu commented:
hmm. the `TestSparkDataFile` failed after this change.
```
Caused by: org.apache.iceberg.exceptions.ValidationException: Invalid schema: multiple fields for name ts: 9 and 9
	at org.apache.iceberg.exceptions.ValidationException.check(ValidationException.java:49)
	at org.apache.iceberg.types.IndexByName.addField(IndexByName.java:195)
	at org.apache.iceberg.types.IndexByName.field(IndexByName.java:161)
	at org.apache.iceberg.types.IndexByName.field(IndexByName.java:34)
	at org.apache.iceberg.types.TypeUtil.visit(TypeUtil.java:611)
	at org.apache.iceberg.types.TypeUtil.indexNameById(TypeUtil.java:173)
	at org.apache.iceberg.Schema.lazyIdToName(Schema.java:213)
	at org.apache.iceberg.Schema.<init>(Schema.java:142)
	at org.apache.iceberg.Schema.<init>(Schema.java:99)
	at org.apache.iceberg.Schema.<init>(Schema.java:87)
	at org.apache.iceberg.Schema.<init>(Schema.java:83)
	at org.apache.iceberg.StructTransform.<init>(StructTransform.java:82)
	at org.apache.iceberg.PartitionKey.<init>(PartitionKey.java:37)
	at org.apache.iceberg.spark.source.SparkWrite$PartitionedDataWriter.<init>(SparkWrite.java:774)
```

This test class apply two partition spec transformation (identify and hour) on the same field `ts`. Is that a legitimate use case? If it is yes, then we have a problem with this proposed change.

```
  private static final PartitionSpec SPEC =
      PartitionSpec.builderFor(SCHEMA)
          .identity("b")
          .bucket("i", 2)
          .identity("l")
          .identity("f")
          .identity("d")
          .identity("date")
          .hour("ts")
          .identity("ts")
          .identity("tsntz")
          .truncate("s", 2)
          .identity("bytes")
          .bucket("dec_9_0", 2)
          .bucket("dec_11_2", 2)
          .bucket("dec_38_10", 2)
          .build();
```

cc @rdblue @aokolnychyi @RussellSpitzer @szehon-ho

RussellSpitzer commented on api/src/main/java/org/apache/iceberg/StructTransform.java:
```diff
@@ -125,6 +151,11 @@ public int hashCode() {
   }
 
+  /** Return the schema of transformed result (a flattened structure) */
+  public Schema resultSchema() {
```
My only question here is why we want to produce this as a Schema? Wouldn't matching the List<Transforms> fit better here? Then we wouldn't have to create artificial fieldId's. If we mostly need Schema out of this api though that would make sense though.

stevenzwu replied:
Do you mean `List<NestedField>` or `List<Type>`? the former still requires field ids. 

For my use case (Flink sink shuffling), we want to use `StructLikeComparator` to compare the `SortKey` values. Hence need the result schema of the `SortKey/StructTransform`.

Admittedly, the field ids and names here are very meaningful. With `List<Type>`, I can construct a schema with artificial field ids and names at the caller side. so it is also workable.

stevenzwu replied:
following on that logic, let me think more if I really need this change. Maybe I can just extract `List<Type> resultTypes` from the `List<FieldTransform> fieldTransforms`

stevenzwu replied:
> Maybe I can just extract List<Type> resultTypes from the List<FieldTransform> fieldTransforms

Yeah, this should be doable. I can construct the schema with artificial field ids and names at the caller side
